File: drivers/apic.c
/*
 * SlopOS APIC (Advanced Programmable Interrupt Controller) Driver
 * Local APIC and I/O APIC detection and basic initialization
 */

#include "apic.h"
#include "serial.h"
#include "../lib/klog.h"
#include "../lib/cpu.h"
#include "../boot/limine_protocol.h"

// Global APIC state
static int apic_available = 0;
static int x2apic_available = 0;
static uint64_t apic_base_address = 0;
static uint64_t apic_base_physical = 0;
static int apic_enabled = 0;

/*
 * Detect APIC availability
 */
int apic_detect(void) {
    uint32_t eax, ebx, ecx, edx;

    klog_debug("APIC: Detecting Local APIC availability");

    // Check CPUID leaf 1 for APIC support
    cpuid(1, &eax, &ebx, &ecx, &edx);

    // Check for Local APIC in EDX bit 9
    if (edx & CPUID_FEAT_EDX_APIC) {
        apic_available = 1;
        klog_debug("APIC: Local APIC is available");

        // Check for x2APIC in ECX bit 21
        if (ecx & CPUID_FEAT_ECX_X2APIC) {
            x2apic_available = 1;
            klog_debug("APIC: x2APIC mode is available");
        } else {
            klog_debug("APIC: x2APIC mode is not available");
        }

        // Get APIC base address from MSR
        uint64_t apic_base_msr = read_msr(MSR_APIC_BASE);
        apic_base_physical = apic_base_msr & APIC_BASE_ADDR_MASK;

        klog_printf(KLOG_DEBUG, "APIC: Physical base: 0x%llx\n",
                    (unsigned long long)apic_base_physical);

        if (is_hhdm_available()) {
            uint64_t hhdm_offset = get_hhdm_offset();
            apic_base_address = apic_base_physical + hhdm_offset;

            klog_printf(KLOG_DEBUG, "APIC: Virtual base (HHDM): 0x%llx\n",
                        (unsigned long long)apic_base_address);
        } else {
            klog_info("APIC: ERROR - HHDM not available, cannot map APIC registers");
            apic_available = 0;
            return 0;
        }

        klog_printf(KLOG_DEBUG, "APIC: MSR flags:%s%s%s\n",
                    (apic_base_msr & APIC_BASE_BSP) ? " BSP" : "",
                    (apic_base_msr & APIC_BASE_X2APIC) ? " X2APIC" : "",
                    (apic_base_msr & APIC_BASE_GLOBAL_ENABLE) ? " ENABLED" : "");

        return 1;
    } else {
        klog_debug("APIC: Local APIC is not available");
        return 0;
    }
}

/*
 * Initialize APIC
 */
int apic_init(void) {
    if (!apic_available) {
        klog_printf(KLOG_INFO, "APIC: Cannot initialize - APIC not available\n");
        return -1;
    }

    klog_debug("APIC: Initializing Local APIC");

    // Enable APIC globally in MSR if not already enabled
    uint64_t apic_base_msr = read_msr(MSR_APIC_BASE);
    if (!(apic_base_msr & APIC_BASE_GLOBAL_ENABLE)) {
        apic_base_msr |= APIC_BASE_GLOBAL_ENABLE;
        write_msr(MSR_APIC_BASE, apic_base_msr);
        klog_debug("APIC: Enabled APIC globally via MSR");
    }

    // Enable APIC via spurious vector register
    apic_enable();

    // Mask all LVT entries to prevent spurious interrupts
    apic_write_register(LAPIC_LVT_TIMER, LAPIC_LVT_MASKED);
    apic_write_register(LAPIC_LVT_LINT0, LAPIC_LVT_MASKED);
    apic_write_register(LAPIC_LVT_LINT1, LAPIC_LVT_MASKED);
    apic_write_register(LAPIC_LVT_ERROR, LAPIC_LVT_MASKED);
    apic_write_register(LAPIC_LVT_PERFCNT, LAPIC_LVT_MASKED);

    /* Route legacy PIC interrupts through LINT0 in ExtINT mode */
    uint32_t lint0 = LAPIC_LVT_DELIVERY_MODE_EXTINT;
    apic_write_register(LAPIC_LVT_LINT0, lint0);

    // Clear error status register
    apic_write_register(LAPIC_ESR, 0);
    apic_write_register(LAPIC_ESR, 0);  // Write twice as per Intel manual

    // Clear any pending EOI
    apic_send_eoi();

    uint32_t apic_id = apic_get_id();
    uint32_t apic_version = apic_get_version();

    klog_printf(KLOG_DEBUG, "APIC: ID: 0x%x, Version: 0x%x\n", apic_id, apic_version);

    apic_enabled = 1;
    klog_debug("APIC: Initialization complete");

    return 0;
}

/*
 * Check if APIC is available
 */
int apic_is_available(void) {
    return apic_available;
}

/*
 * Check if x2APIC is available
 */
int apic_is_x2apic_available(void) {
    return x2apic_available;
}

/*
 * Check if this is the Bootstrap Processor
 */
int apic_is_bsp(void) {
    if (!apic_available) return 0;
    uint64_t apic_base_msr = read_msr(MSR_APIC_BASE);
    return (apic_base_msr & APIC_BASE_BSP) != 0;
}

int apic_is_enabled(void) {
    return apic_enabled;
}

/*
 * Enable APIC
 */
void apic_enable(void) {
    if (!apic_available) return;

    // Enable APIC via spurious vector register
    uint32_t spurious = apic_read_register(LAPIC_SPURIOUS);
    spurious |= LAPIC_SPURIOUS_ENABLE;
    spurious |= 0xFF;  // Set spurious vector to 255
    apic_write_register(LAPIC_SPURIOUS, spurious);

    apic_enabled = 1;
    klog_debug("APIC: Local APIC enabled");
}

/*
 * Disable APIC
 */
void apic_disable(void) {
    if (!apic_available) return;

    // Disable APIC via spurious vector register
    uint32_t spurious = apic_read_register(LAPIC_SPURIOUS);
    spurious &= ~LAPIC_SPURIOUS_ENABLE;
    apic_write_register(LAPIC_SPURIOUS, spurious);

    apic_enabled = 0;
    klog_debug("APIC: Local APIC disabled");
}

/*
 * Send End of Interrupt
 */
void apic_send_eoi(void) {
    if (!apic_enabled) return;
    apic_write_register(LAPIC_EOI, 0);
}

/*
 * Get APIC ID
 */
uint32_t apic_get_id(void) {
    if (!apic_available) return 0;
    uint32_t id = apic_read_register(LAPIC_ID);
    return id >> 24;  // APIC ID is in bits 31:24
}

/*
 * Get APIC version
 */
uint32_t apic_get_version(void) {
    if (!apic_available) return 0;
    return apic_read_register(LAPIC_VERSION) & 0xFF;
}

/*
 * Initialize APIC timer
 */
void apic_timer_init(uint32_t vector, uint32_t frequency) {
    if (!apic_enabled) return;

    klog_printf(KLOG_DEBUG, "APIC: Initializing timer with vector 0x%x and frequency %u\n",
                vector, frequency);

    // Set timer divisor to 16
    apic_timer_set_divisor(LAPIC_TIMER_DIV_16);

    // Configure timer LVT
    uint32_t lvt_timer = vector | LAPIC_TIMER_PERIODIC;
    apic_write_register(LAPIC_LVT_TIMER, lvt_timer);

    // Calculate initial count for desired frequency
    // This is a rough calculation - would need calibration for accuracy
    uint32_t initial_count = 1000000 / frequency;  // Approximate
    apic_timer_start(initial_count);

    klog_debug("APIC: Timer initialized");
}

/*
 * Start APIC timer
 */
void apic_timer_start(uint32_t initial_count) {
    if (!apic_enabled) return;
    apic_write_register(LAPIC_TIMER_ICR, initial_count);
}

/*
 * Stop APIC timer
 */
void apic_timer_stop(void) {
    if (!apic_enabled) return;
    apic_write_register(LAPIC_TIMER_ICR, 0);
}

/*
 * Get current timer count
 */
uint32_t apic_timer_get_current_count(void) {
    if (!apic_enabled) return 0;
    return apic_read_register(LAPIC_TIMER_CCR);
}

/*
 * Set timer divisor
 */
void apic_timer_set_divisor(uint32_t divisor) {
    if (!apic_enabled) return;
    apic_write_register(LAPIC_TIMER_DCR, divisor);
}

/*
 * Get APIC base address
 */
uint64_t apic_get_base_address(void) {
    return apic_base_address;
}

/*
 * Set APIC base address
 */
void apic_set_base_address(uint64_t base) {
    if (!apic_available) return;

    uint64_t masked_base = base & APIC_BASE_ADDR_MASK;
    uint64_t apic_base_msr = read_msr(MSR_APIC_BASE);
    apic_base_msr = (apic_base_msr & ~APIC_BASE_ADDR_MASK) | masked_base;
    write_msr(MSR_APIC_BASE, apic_base_msr);

    apic_base_physical = masked_base;
    if (is_hhdm_available()) {
        apic_base_address = apic_base_physical + get_hhdm_offset();
    } else {
        apic_base_address = 0;
    }
}

/*
 * Read APIC register
 */
uint32_t apic_read_register(uint32_t reg) {
    if (!apic_available || apic_base_address == 0) return 0;

    // Memory-mapped access to APIC registers
    volatile uint32_t *apic_reg = (volatile uint32_t *)(uintptr_t)(apic_base_address + reg);
    return *apic_reg;
}

/*
 * Write APIC register
 */
void apic_write_register(uint32_t reg, uint32_t value) {
    if (!apic_available || apic_base_address == 0) return;

    // Memory-mapped access to APIC registers
    volatile uint32_t *apic_reg = (volatile uint32_t *)(uintptr_t)(apic_base_address + reg);
    *apic_reg = value;
}

/*
 * Dump APIC state for debugging
 */
void apic_dump_state(void) {
    klog_printf(KLOG_INFO, "=== APIC STATE DUMP ===\n");

    if (!apic_available) {
        klog_printf(KLOG_INFO, "APIC: Not available\n");
        klog_printf(KLOG_INFO, "=== END APIC STATE DUMP ===\n");
        return;
    }

    klog_printf(KLOG_INFO, "APIC Available: Yes, x2APIC: %s\n", x2apic_available ? "Yes" : "No");
    klog_printf(KLOG_INFO, "APIC Enabled: %s\n", apic_enabled ? "Yes" : "No");
    klog_printf(KLOG_INFO, "Bootstrap Processor: %s\n", apic_is_bsp() ? "Yes" : "No");
    klog_printf(KLOG_INFO, "Base Address: 0x%llx\n", (unsigned long long)apic_base_address);

    if (apic_enabled) {
        klog_printf(KLOG_INFO, "APIC ID: 0x%x\n", apic_get_id());
        klog_printf(KLOG_INFO, "APIC Version: 0x%x\n", apic_get_version());

        uint32_t spurious = apic_read_register(LAPIC_SPURIOUS);
        klog_printf(KLOG_INFO, "Spurious Vector Register: 0x%x\n", spurious);

        uint32_t esr = apic_read_register(LAPIC_ESR);
        klog_printf(KLOG_INFO, "Error Status Register: 0x%x\n", esr);

        uint32_t lvt_timer = apic_read_register(LAPIC_LVT_TIMER);
        klog_printf(KLOG_INFO, "Timer LVT: 0x%x%s\n",
                    lvt_timer,
                    (lvt_timer & LAPIC_LVT_MASKED) ? " (MASKED)" : "");

        uint32_t timer_count = apic_timer_get_current_count();
        klog_printf(KLOG_INFO, "Timer Current Count: 0x%x\n", timer_count);
    }

    klog_printf(KLOG_INFO, "=== END APIC STATE DUMP ===\n");
}

File: drivers/src/apic.rs
use core::ptr::{read_volatile, write_volatile};
use core::sync::atomic::{AtomicBool, AtomicU64, Ordering};

use slopos_lib::{cpu, klog_debug, klog_info};

use crate::hw::apic_defs::*;
use crate::sched_bridge;
use crate::wl_currency;

static APIC_AVAILABLE: AtomicBool = AtomicBool::new(false);
static X2APIC_AVAILABLE: AtomicBool = AtomicBool::new(false);
static APIC_ENABLED: AtomicBool = AtomicBool::new(false);
static APIC_BASE_ADDRESS: AtomicU64 = AtomicU64::new(0);
static APIC_BASE_PHYSICAL: AtomicU64 = AtomicU64::new(0);

#[inline]
fn hhdm_virt_for(phys: u64) -> Option<u64> {
    if phys == 0 {
        return None;
    }
    if sched_bridge::is_hhdm_available() != 0 {
        Some(phys + sched_bridge::get_hhdm_offset())
    } else {
        None
    }
}

pub fn detect() -> bool {
    klog_debug!("APIC: Detecting Local APIC availability...");

    let (_, _, ecx, edx) = cpu::cpuid(1);
    if edx & CPUID_FEAT_EDX_APIC == 0 {
        klog_debug!("APIC: Local APIC is not available");
        APIC_AVAILABLE.store(false, Ordering::Relaxed);
        return false;
    }

    APIC_AVAILABLE.store(true, Ordering::Relaxed);
    let x2 = (ecx & CPUID_FEAT_ECX_X2APIC) != 0;
    X2APIC_AVAILABLE.store(x2, Ordering::Relaxed);

    let apic_base_msr = cpu::read_msr(MSR_APIC_BASE);
    let apic_phys = apic_base_msr & APIC_BASE_ADDR_MASK;
    APIC_BASE_PHYSICAL.store(apic_phys, Ordering::Relaxed);

    if let Some(virt) = hhdm_virt_for(apic_phys) {
        APIC_BASE_ADDRESS.store(virt, Ordering::Relaxed);
        let bsp_flag = if apic_base_msr & APIC_BASE_BSP != 0 {
            " BSP"
        } else {
            ""
        };
        let x2apic_flag = if apic_base_msr & APIC_BASE_X2APIC != 0 {
            " X2APIC"
        } else {
            ""
        };
        let enable_flag = if apic_base_msr & APIC_BASE_GLOBAL_ENABLE != 0 {
            " ENABLED"
        } else {
            ""
        };
        klog_debug!(
            "APIC: Physical base: 0x{:x}, Virtual base (HHDM): 0x{:x}",
            apic_phys,
            virt
        );
        klog_debug!("APIC: MSR flags:{}{}{}", bsp_flag, x2apic_flag, enable_flag);
        true
    } else {
        klog_info!("APIC: ERROR - HHDM not available, cannot map APIC registers");
        APIC_AVAILABLE.store(false, Ordering::Relaxed);
        false
    }
}

pub fn init() -> i32 {
    if !is_available() {
        klog_info!("APIC: Cannot initialize - APIC not available");
        wl_currency::award_loss();
        return -1;
    }

    klog_debug!("APIC: Initializing Local APIC");

    let mut apic_base_msr = cpu::read_msr(MSR_APIC_BASE);
    if apic_base_msr & APIC_BASE_GLOBAL_ENABLE == 0 {
        apic_base_msr |= APIC_BASE_GLOBAL_ENABLE;
        cpu::write_msr(MSR_APIC_BASE, apic_base_msr);
        klog_debug!("APIC: Enabled APIC globally via MSR");
    }

    enable();

    // Mask all LVT entries to prevent spurious interrupts.
    write_register(LAPIC_LVT_TIMER, LAPIC_LVT_MASKED);
    write_register(LAPIC_LVT_LINT0, LAPIC_LVT_MASKED);
    write_register(LAPIC_LVT_LINT1, LAPIC_LVT_MASKED);
    write_register(LAPIC_LVT_ERROR, LAPIC_LVT_MASKED);
    write_register(LAPIC_LVT_PERFCNT, LAPIC_LVT_MASKED);

    // Route legacy PIC interrupts through LINT0 in ExtINT mode.
    write_register(LAPIC_LVT_LINT0, LAPIC_LVT_DELIVERY_MODE_EXTINT);

    // Clear error status register twice per Intel SDM guidance.
    write_register(LAPIC_ESR, 0);
    write_register(LAPIC_ESR, 0);

    send_eoi();

    let apic_id = get_id();
    let apic_version = get_version();
    klog_debug!("APIC: ID: 0x{:x}, Version: 0x{:x}", apic_id, apic_version);

    APIC_ENABLED.store(true, Ordering::Relaxed);
    klog_debug!("APIC: Initialization complete");
    wl_currency::award_win();
    0
}

pub fn is_available() -> bool {
    APIC_AVAILABLE.load(Ordering::Relaxed)
}

pub fn is_x2apic_available() -> bool {
    X2APIC_AVAILABLE.load(Ordering::Relaxed)
}

pub fn is_bsp() -> bool {
    if !is_available() {
        return false;
    }
    let apic_base_msr = cpu::read_msr(MSR_APIC_BASE);
    (apic_base_msr & APIC_BASE_BSP) != 0
}

pub fn is_enabled() -> bool {
    APIC_ENABLED.load(Ordering::Relaxed)
}

pub fn enable() {
    if !is_available() {
        return;
    }
    let mut spurious = read_register(LAPIC_SPURIOUS);
    spurious |= LAPIC_SPURIOUS_ENABLE;
    spurious |= 0xFF;
    write_register(LAPIC_SPURIOUS, spurious);
    APIC_ENABLED.store(true, Ordering::Relaxed);
    klog_debug!("APIC: Local APIC enabled");
}

pub fn disable() {
    if !is_available() {
        return;
    }
    let mut spurious = read_register(LAPIC_SPURIOUS);
    spurious &= !LAPIC_SPURIOUS_ENABLE;
    write_register(LAPIC_SPURIOUS, spurious);
    APIC_ENABLED.store(false, Ordering::Relaxed);
    klog_debug!("APIC: Local APIC disabled");
}

pub fn send_eoi() {
    if !is_enabled() {
        return;
    }
    write_register(LAPIC_EOI, 0);
}

pub fn get_id() -> u32 {
    if !is_available() {
        return 0;
    }
    read_register(LAPIC_ID) >> 24
}

pub fn get_version() -> u32 {
    if !is_available() {
        return 0;
    }
    read_register(LAPIC_VERSION) & 0xFF
}

pub fn timer_init(vector: u32, frequency: u32) {
    if !is_enabled() {
        return;
    }
    klog_debug!(
        "APIC: Initializing timer with vector 0x{:x} and frequency {}",
        vector,
        frequency
    );

    timer_set_divisor(LAPIC_TIMER_DIV_16);

    let lvt_timer = vector | LAPIC_TIMER_PERIODIC;
    write_register(LAPIC_LVT_TIMER, lvt_timer);

    let initial_count = 1_000_000u32.saturating_div(frequency.max(1));
    timer_start(initial_count);
    klog_debug!("APIC: Timer initialized");
}

pub fn timer_start(initial_count: u32) {
    if !is_enabled() {
        return;
    }
    write_register(LAPIC_TIMER_ICR, initial_count);
}

pub fn timer_stop() {
    if !is_enabled() {
        return;
    }
    write_register(LAPIC_TIMER_ICR, 0);
}

pub fn timer_get_current_count() -> u32 {
    if !is_enabled() {
        return 0;
    }
    read_register(LAPIC_TIMER_CCR)
}

pub fn timer_set_divisor(divisor: u32) {
    if !is_enabled() {
        return;
    }
    write_register(LAPIC_TIMER_DCR, divisor);
}

pub fn send_ipi_halt_all() {
    if !is_available() || !is_enabled() {
        return;
    }

    // Vector 0xFE is commonly used for shutdown IPI
    const SHUTDOWN_VECTOR: u32 = 0xFE;

    // Wait for any pending IPI to complete (delivery status bit must be clear)
    let mut timeout = 1000;
    while (read_register(LAPIC_ICR_LOW) & LAPIC_ICR_DELIVERY_STATUS) != 0 && timeout > 0 {
        cpu::pause();
        timeout -= 1;
    }

    // Write destination to ICR_HIGH (broadcast to all except self in physical mode)
    write_register(LAPIC_ICR_HIGH, LAPIC_ICR_DEST_BROADCAST);

    // Write IPI command to ICR_LOW:
    // - Vector: 0xFE (shutdown)
    // - Delivery mode: Fixed (0)
    // - Destination mode: Physical (0)
    // - Level: Asserted (1)
    // - Trigger: Edge (0)
    let icr_low = SHUTDOWN_VECTOR
        | LAPIC_ICR_DELIVERY_FIXED
        | LAPIC_ICR_DEST_PHYSICAL
        | LAPIC_ICR_LEVEL_ASSERT
        | LAPIC_ICR_TRIGGER_EDGE;
    write_register(LAPIC_ICR_LOW, icr_low);

    // Wait for IPI to be sent (delivery status bit must clear)
    timeout = 1000;
    while (read_register(LAPIC_ICR_LOW) & LAPIC_ICR_DELIVERY_STATUS) != 0 && timeout > 0 {
        cpu::pause();
        timeout -= 1;
    }

    klog_debug!("APIC: Sent shutdown IPI to all processors");
}

pub fn get_base_address() -> u64 {
    APIC_BASE_ADDRESS.load(Ordering::Relaxed)
}

pub fn set_base_address(base: u64) {
    if !is_available() {
        return;
    }
    let masked_base = base & APIC_BASE_ADDR_MASK;
    let mut apic_base_msr = cpu::read_msr(MSR_APIC_BASE);
    apic_base_msr = (apic_base_msr & !APIC_BASE_ADDR_MASK) | masked_base;
    cpu::write_msr(MSR_APIC_BASE, apic_base_msr);

    APIC_BASE_PHYSICAL.store(masked_base, Ordering::Relaxed);
    if let Some(virt) = hhdm_virt_for(masked_base) {
        APIC_BASE_ADDRESS.store(virt, Ordering::Relaxed);
    } else {
        APIC_BASE_ADDRESS.store(0, Ordering::Relaxed);
    }
}

pub fn read_register(reg: u32) -> u32 {
    let base = APIC_BASE_ADDRESS.load(Ordering::Relaxed);
    if !is_available() || base == 0 {
        return 0;
    }
    let reg_ptr = (base + reg as u64) as *const u32;
    unsafe { read_volatile(reg_ptr) }
}

pub fn write_register(reg: u32, value: u32) {
    let base = APIC_BASE_ADDRESS.load(Ordering::Relaxed);
    if !is_available() || base == 0 {
        return;
    }
    let reg_ptr = (base + reg as u64) as *mut u32;
    unsafe { write_volatile(reg_ptr, value) };
}

pub fn dump_state() {
    klog_info!("=== APIC STATE DUMP ===");
    if !is_available() {
        klog_info!("APIC: Not available");
        klog_info!("=== END APIC STATE DUMP ===");
        return;
    }

    klog_info!(
        "APIC Available: Yes, x2APIC: {}",
        if is_x2apic_available() { "Yes" } else { "No" }
    );
    klog_info!("APIC Enabled: {}", if is_enabled() { "Yes" } else { "No" });
    klog_info!(
        "Bootstrap Processor: {}",
        if is_bsp() { "Yes" } else { "No" }
    );
    klog_info!("Base Address: 0x{:x}", get_base_address());

    if is_enabled() {
        let spurious = read_register(LAPIC_SPURIOUS);
        let esr = read_register(LAPIC_ESR);
        let lvt_timer = read_register(LAPIC_LVT_TIMER);
        let timer_count = timer_get_current_count();
        klog_info!("APIC ID: 0x{:x}", get_id());
        klog_info!("APIC Version: 0x{:x}", get_version());
        klog_info!("Spurious Vector Register: 0x{:x}", spurious);
        klog_info!("Error Status Register: 0x{:x}", esr);
        klog_info!(
            "Timer LVT: 0x{:x}{}",
            lvt_timer,
            if lvt_timer & LAPIC_LVT_MASKED != 0 {
                " (MASKED)"
            } else {
                ""
            }
        );
        klog_info!("Timer Current Count: 0x{:x}", timer_count);
    }

    klog_info!("=== END APIC STATE DUMP ===");
}
pub fn apic_detect() -> i32 {
    if detect() { 1 } else { 0 }
}
pub fn apic_init() -> i32 {
    init()
}
pub fn apic_is_available() -> i32 {
    if is_available() { 1 } else { 0 }
}
pub fn apic_is_x2apic_available() -> i32 {
    if is_x2apic_available() { 1 } else { 0 }
}
pub fn apic_is_bsp() -> i32 {
    if is_bsp() { 1 } else { 0 }
}
pub fn apic_is_enabled() -> i32 {
    if is_enabled() { 1 } else { 0 }
}
pub fn apic_enable() {
    enable();
}
pub fn apic_disable() {
    disable();
}
pub fn apic_send_eoi() {
    send_eoi();
}
pub fn apic_get_id() -> u32 {
    get_id()
}
pub fn apic_get_version() -> u32 {
    get_version()
}
pub fn apic_timer_init(vector: u32, frequency: u32) {
    timer_init(vector, frequency);
}
pub fn apic_timer_start(initial_count: u32) {
    timer_start(initial_count);
}
pub fn apic_timer_stop() {
    timer_stop();
}
pub fn apic_timer_get_current_count() -> u32 {
    timer_get_current_count()
}
pub fn apic_timer_set_divisor(divisor: u32) {
    timer_set_divisor(divisor);
}
pub fn apic_dump_state() {
    dump_state();
}
pub fn apic_get_base_address() -> u64 {
    get_base_address()
}
pub fn apic_set_base_address(base: u64) {
    set_base_address(base);
}
pub fn apic_read_register(reg: u32) -> u32 {
    read_register(reg)
}
pub fn apic_write_register(reg: u32, value: u32) {
    write_register(reg, value);
}
pub fn apic_send_ipi_halt_all() {
    send_ipi_halt_all();
}
